File: storms.py
import random
import numpy as np

from globals import PLANET_RADIUS_KM
from utils import calculate_bearing, cartesian_to_lat_lon, find_spherical_neighbors, haversine_distance


class StormSystem:
    def __init__(self, center_lat, center_lon, pressure_anomaly, radius_km=500, movement_speed_km_day=50):
        self.center_lat = center_lat
        self.center_lon = center_lon
        self.pressure_anomaly = pressure_anomaly  # Negative for cyclones, positive for anticyclones
        self.radius_km = radius_km
        self.movement_speed_km_day = movement_speed_km_day
        self.movement_direction = np.random.uniform(0, 360)  # Degrees from north
        self.age_days = 0
        self.max_lifetime_days = np.random.uniform(5, 15)  # Storm lifespan
        self.intensity_change_rate = np.random.uniform(-0.05, 0.05)  # Daily intensity change
        
    def update(self, vertices, elevations, temperatures, pressures, faces, day_elapsed=1):
        # Use temperature gradient to steer storm
        nearest_vertex = self._find_nearest_vertex(vertices)
        temp = temperatures[nearest_vertex]
        
        # Move toward warmer areas (for tropical systems)
        if abs(self.center_lat) < 30 and temp > 26:
            self.movement_direction = self._calculate_steering_flow(pressures, vertices, faces)
        
        # Intensify over warm water
        if elevations[nearest_vertex] < 0 and temp > 26:
            self.pressure_anomaly *= 1.02  # 2% intensification per day
        
        # Weaken over land
        elif elevations[nearest_vertex] > 0:
            self.pressure_anomaly *= 0.95  # 5% weakening per day
    
    def _find_nearest_vertex(self, vertices):
        # Convert all vertices to lat/lon first
        vertex_coords = np.array([cartesian_to_lat_lon(*v) for v in vertices])
        vertex_lats = vertex_coords[:, 0]
        vertex_lons = vertex_coords[:, 1]
        
        # Calculate all distances at once
        distances = haversine_distance(
            self.center_lat, self.center_lon,
            vertex_lats, vertex_lons
        )
        
        return np.argmin(distances)
            
    def calculate_pressure_effect(self, lat, lon):
        """Calculate pressure effect at given coordinates."""
        distance = haversine_distance(self.center_lat, self.center_lon, lat, lon, PLANET_RADIUS_KM)
        
        if distance > self.radius_km:
            return 0
            
        # Gaussian influence
        influence = np.exp(-(distance**2)/(2*(self.radius_km/3)**2))
        return self.pressure_anomaly * influence
    
    def calculate_wind_field(self, lat, lon):
        """Calculate wind vector at given coordinates"""
        distance = haversine_distance(self.center_lat, self.center_lon, lat, lon)
        
        # Add cyclonic rotation (counter-clockwise in NH, clockwise in SH)
        bearing = calculate_bearing(self.center_lat, self.center_lon, lat, lon)
        if distance < self.radius_km:
            rotation_dir = 1 if self.center_lat >= 0 else -1  # NH vs SH
            wind_dir = (bearing + 90 * rotation_dir) % 360
            
            # Wind speed based on pressure gradient and distance from center
            max_speed = -self.pressure_anomaly * 3  # Convert pressure to wind speed
            speed = max_speed * (distance/self.radius_km) * np.exp(1 - (distance/self.radius_km))
            
            return wind_dir, speed
        return None
    
    def is_in_rain_band(self, lat, lon):
        """Check if coordinates are within the storm's rain bands"""
        distance = haversine_distance(self.center_lat, self.center_lon, lat, lon)
        
        # Rain bands extend 1.5x the radius of the pressure anomaly
        if distance > self.radius_km * 1.5:
            return False
            
        # Spiral rain band pattern
        angle = np.degrees(np.arctan2(lon - self.center_lon, lat - self.center_lat))
        band_width = 30  # degrees
        return (distance % (self.radius_km/3)) < (self.radius_km/10)
    
    def get_rainfall_intensity(self, lat, lon):
        """Calculate rainfall intensity based on distance from storm center"""
        distance = haversine_distance(self.center_lat, self.center_lon, lat, lon)
        
        if distance > self.radius_km * 1.5:
            return 0
            
        # Max rainfall near the eyewall (0.8-1.2x radius)
        if 0.8 * self.radius_km < distance < 1.2 * self.radius_km:
            return min(300, abs(self.pressure_anomaly) * 5)  # Cap at 300mm
            
        # Decrease with distance
        return max(0, abs(self.pressure_anomaly) * 3 * (1 - distance/(self.radius_km*1.5)))
    
    def _calculate_steering_flow(self, pressures, vertices, faces):
        # Find nearest vertex
        nearest_idx = self._find_nearest_vertex(vertices)
        lat, lon = cartesian_to_lat_lon(*vertices[nearest_idx])
        
        # Find neighboring vertices within 500km
        neighbors = find_spherical_neighbors(vertices, faces, nearest_idx, 500)
        
        if not neighbors:
            return self.movement_direction  # No change if no neighbors
        
        # Calculate pressure gradient
        neighbor_pressures = [pressures[n] for n in neighbors]
        avg_pressure = np.mean(neighbor_pressures)
        pressure_diff = pressures[nearest_idx] - avg_pressure
        
        # Get coordinates of neighbors
        neighbor_coords = np.array([cartesian_to_lat_lon(*vertices[n]) for n in neighbors])
        neighbor_lats = neighbor_coords[:, 0]
        neighbor_lons = neighbor_coords[:, 1]
        
        # Calculate centroid of lower pressure area
        if pressure_diff > 0:  # We're in higher pressure - move toward lower
            low_pressure_neighbors = [n for n in neighbors if pressures[n] < avg_pressure]
            if low_pressure_neighbors:
                target_lat = np.mean([cartesian_to_lat_lon(*vertices[n])[0] for n in low_pressure_neighbors])
                target_lon = np.mean([cartesian_to_lat_lon(*vertices[n])[1] for n in low_pressure_neighbors])
            else:
                return self.movement_direction
        else:  # Already in low pressure - move down gradient
            pressure_grad_x = np.mean(np.sin(np.radians(neighbor_lons)) * (pressures[neighbors] - pressures[nearest_idx]))
            pressure_grad_y = np.mean(np.cos(np.radians(neighbor_lats)) * (pressures[neighbors] - pressures[nearest_idx]))
            
            # Move perpendicular to gradient (geostrophic flow)
            if self.center_lat >= 0:  # Northern hemisphere - counterclockwise
                target_lon = lon + pressure_grad_y * 100
                target_lat = lat - pressure_grad_x * 100
            else:  # Southern hemisphere - clockwise
                target_lon = lon - pressure_grad_y * 100
                target_lat = lat + pressure_grad_x * 100
        
        # Calculate bearing to target
        new_direction = calculate_bearing(lat, lon, target_lat, target_lon)
        
        # Smooth direction change (max 15 degrees per day)
        direction_change = (new_direction - self.movement_direction + 360) % 360
        if direction_change > 180:
            direction_change -= 360
        
        max_change = 15 * (self.radius_km/500)  # Larger storms turn slower
        actual_change = np.clip(direction_change, -max_change, max_change)
        
        return (self.movement_direction + actual_change) % 360


def generate_storm_systems(vertices, elevations, day_of_year, num_storms=5):
    """Generate new storm systems based on current conditions."""
    storms = []
    
    # More storms during seasonal extremes
    seasonal_factor = 1 + 0.5 * np.sin(2 * np.pi * (day_of_year - 80) / 365.25)
    
    for _ in range(int(num_storms * seasonal_factor)):
        # Prefer certain latitudes for storm formation
        if random.random() < 0.7:  # 70% chance in mid-latitudes
            lat = np.random.uniform(-60, 60)
            if abs(lat) < 30:  # Tropical storms
                pressure_anomaly = np.random.uniform(-15, -5)  # Strong low pressure
                radius = np.random.uniform(300, 800)
                speed = np.random.uniform(30, 60)
            else:  # Extratropical cyclones
                pressure_anomaly = np.random.uniform(-25, -10)
                radius = np.random.uniform(500, 1500)
                speed = np.random.uniform(50, 100)
        else:  # 30% chance elsewhere
            lat = np.random.uniform(-90, 90)
            pressure_anomaly = np.random.uniform(-10, 10)
            radius = np.random.uniform(200, 1000)
            speed = np.random.uniform(20, 80)
            
        lon = np.random.uniform(-180, 180)
        
        # Prefer forming over water
        nearest_vertex = min(range(len(vertices)), 
                           key=lambda i: haversine_distance(lat, lon, *cartesian_to_lat_lon(*vertices[i])))
        if elevations[nearest_vertex] < 0:  # Water
            pressure_anomaly *= 1.5  # Stronger over water
            
        storms.append(StormSystem(lat, lon, pressure_anomaly, radius, speed))
        
    return storms

File: storms.cpp
#include "storms.h"
#include <algorithm>
#include <numeric>
#include "utils.cpp"  // Assuming you have C++ utils for geographic calculations

StormSystem::StormSystem(double lat, double lon, double anomaly, double radius, double speed) :
    center_lat(lat), center_lon(lon), pressure_anomaly(anomaly), 
    radius_km(radius), movement_speed_km_day(speed) {
    
    static std::random_device rd;
    static std::mt19937 gen(rd());
    std::uniform_real_distribution<> dir_dist(0.0, 360.0);
    std::uniform_real_distribution<> life_dist(5.0, 15.0);
    std::uniform_real_distribution<> int_dist(-0.05, 0.05);
    
    movement_direction = dir_dist(gen);
    max_lifetime_days = life_dist(gen);
    intensity_change_rate = int_dist(gen);
    age_days = 0;
}

void StormSystem::update(const std::vector<std::vector<double>>& vertices, const std::vector<double>& elevations,
                       const std::vector<double>& temperatures, const std::vector<double>& pressures,
                       const std::vector<std::vector<int>>& faces,
                       int day_elapsed) {
    
    int nearest_vertex = find_nearest_vertex(vertices);
    double temp = temperatures[nearest_vertex];
    
    // Move toward warmer areas (for tropical systems)
    if (std::abs(center_lat) < 30.0 && temp > 26.0) {
        movement_direction = calculate_steering_flow(pressures, vertices, faces);
    }
    
    // Intensify over warm water
    if (elevations[nearest_vertex] < 0.0 && temp > 26.0) {
        pressure_anomaly *= 1.02;  // 2% intensification per day
    } else if (elevations[nearest_vertex] > 0.0) {  // Weaken over land
        pressure_anomaly *= 0.95;  // 5% weakening per day
    }
    
    age_days += day_elapsed;
}

int StormSystem::find_nearest_vertex(const vector<vector<Vertex>>& vertices) {
    int nearest = 0;
    double min_dist = std::numeric_limits<double>::max();
    
    for (int i = 0; i < vertices.size(); ++i) {
        pair<double, double> latlon = cartesian_to_lat_lon(vertices[i].x, vertices[i].y, vertices[i].z);
        double dist = haversine_distance(center_lat, center_lon, latlon.first, latlon.second);
        if (dist < min_dist) {
            min_dist = dist;
            nearest = i;
        }
    }
    return nearest;
}

double StormSystem::calculate_pressure_effect(double lat, double lon, double planet_radius_km) const {
    double distance = haversine_distance(center_lat, center_lon, lat, lon, planet_radius_km);
    
    if (distance > radius_km) {
        return 0.0;
    }
    
    // Gaussian influence
    double influence = std::exp(-(distance * distance) / (2.0 * std::pow(radius_km / 3.0, 2)));
    return pressure_anomaly * influence;
}

std::pair<double, double> StormSystem::calculate_wind_field(double lat, double lon) const {
    double storm_distance = haversine_distance(center_lat, center_lon, lat, lon);
    
    if (storm_distance < radius_km) {
        double bearing = calculate_bearing_math(center_lat, center_lon, lat, lon);
        int rotation_dir = (center_lat >= 0) ? 1 : -1;  // NH vs SH
        double wind_dir = fmod(bearing + 90.0 * rotation_dir, 360.0);
        
        // Wind speed based on pressure gradient and distance from center
        double max_speed = -pressure_anomaly * 3.0;  // Convert pressure to wind speed
        double speed = max_speed * (storm_distance / radius_km) * std::exp(1.0 - (storm_distance / radius_km));
        
        return {wind_dir, speed};
    }
    return {0.0, 0.0};  // No wind outside storm radius
}

bool StormSystem::is_in_rain_band(double lat, double lon) const {
    double distance = haversine_distance(center_lat, center_lon, lat, lon);
    
    // Rain bands extend 1.5x the radius of the pressure anomaly
    if (distance > radius_km * 1.5) {
        return false;
    }
    
    // Spiral rain band pattern
    double angle = std::atan2(lon - center_lon, lat - center_lat) * 180.0 / PI;
    double band_width = 30.0;  // degrees
    return fmod(distance, radius_km / 3.0) < (radius_km / 10.0);
}

double StormSystem::get_rainfall_intensity(double lat, double lon) const {
    double distance = haversine_distance(center_lat, center_lon, lat, lon);
    
    if (distance > radius_km * 1.5) {
        return 0.0;
    }
    
    // Max rainfall near the eyewall (0.8-1.2x radius)
    if (0.8 * radius_km < distance && distance < 1.2 * radius_km) {
        return std::min(300.0, std::abs(pressure_anomaly) * 5.0);  // Cap at 300mm
    }
    
    // Decrease with distance
    return std::max(0.0, std::abs(pressure_anomaly) * 3.0 * (1.0 - distance / (radius_km * 1.5)));
}

double StormSystem::calculate_steering_flow(const std::vector<double>& pressures, 
                                          const std::vector<std::vector<Vertex>>& vertices,
                                          const std::vector<std::vector<Face>>& faces) {
    int nearest_idx = find_nearest_vertex(vertices);
    auto latlon = cartesian_to_lat_lon(vertices[nearest_idx][0], vertices[nearest_idx][1], vertices[nearest_idx][2]);
    double lat = latlon.first;
    double lon = latlon.second;
    
    // Find neighboring vertices within 500km
    vector<Vertex>& neighbors = find_spherical_neighbors(vertices, faces, nearest_idx, 500.0);
    
    if (neighbors.empty()) {
        return movement_direction;  // No change if no neighbors
    }
    
    // Calculate pressure gradient
    double sum_pressure = 0.0;
    for (int n : neighbors) {
        sum_pressure += pressures[n];
    }
    double avg_pressure = sum_pressure / neighbors.size();
    double pressure_diff = pressures[nearest_idx] - avg_pressure;
    
    double target_lat, target_lon;
    
    if (pressure_diff > 0.0) {  // We're in higher pressure - move toward lower
        std::vector<int> low_pressure_neighbors;
        for (int n : neighbors) {
            if (pressures[n] < avg_pressure) {
                low_pressure_neighbors.push_back(n);
            }
        }
        
        if (!low_pressure_neighbors.empty()) {
            double sum_lat = 0.0, sum_lon = 0.0;
            for (int n : low_pressure_neighbors) {
                auto ll = cartesian_to_lat_lon(vertices[n][0], vertices[n][1], vertices[n][2]);
                sum_lat += ll.first;
                sum_lon += ll.second;
            }
            target_lat = sum_lat / low_pressure_neighbors.size();
            target_lon = sum_lon / low_pressure_neighbors.size();
        } else {
            return movement_direction;
        }
    } else {  // Already in low pressure - move down gradient
        double sum_sin_lon_pressure = 0.0;
        double sum_cos_lat_pressure = 0.0;
        
        for (int n : neighbors) {
            auto ll = cartesian_to_lat_lon(vertices[n][0], vertices[n][1], vertices[n][2]);
            sum_sin_lon_pressure += std::sin(ll.second * PI / 180.0) * (pressures[n] - pressures[nearest_idx]);
            sum_cos_lat_pressure += std::cos(ll.first * PI / 180.0) * (pressures[n] - pressures[nearest_idx]);
        }
        
        double pressure_grad_x = sum_sin_lon_pressure / neighbors.size();
        double pressure_grad_y = sum_cos_lat_pressure / neighbors.size();
        
        // Move perpendicular to gradient (geostrophic flow)
        if (center_lat >= 0.0) {  // Northern hemisphere - counterclockwise
            target_lon = lon + pressure_grad_y * 100.0;
            target_lat = lat - pressure_grad_x * 100.0;
        } else {  // Southern hemisphere - clockwise
            target_lon = lon - pressure_grad_y * 100.0;
            target_lat = lat + pressure_grad_x * 100.0;
        }
    }
    
    // Calculate bearing to target
    double new_direction = calculate_bearing(lat, lon, target_lat, target_lon);
    
    // Smooth direction change (max 15 degrees per day)
    double direction_change = fmod(new_direction - movement_direction + 360.0, 360.0);
    if (direction_change > 180.0) {
        direction_change -= 360.0;
    }
    
    double max_change = 15.0 * (radius_km / 500.0);  // Larger storms turn slower
    direction_change = std::max(-max_change, std::min(max_change, direction_change));
    
    return fmod(movement_direction + direction_change, 360.0);
}

std::vector<StormSystem> generate_storm_systems(const std::vector<std::vector<double>>& vertices,
                                              const std::vector<double>& elevations,
                                              int day_of_year,
                                              int num_storms) {
    std::vector<StormSystem> storms;
    static std::random_device rd;
    static std::mt19937 gen(rd());
    
    // More storms during seasonal extremes
    double seasonal_factor = 1.0 + 0.5 * std::sin(2.0 * PI * (day_of_year - 80) / 365.25);
    int num_to_generate = static_cast<int>(num_storms * seasonal_factor);
    
    std::uniform_real_distribution<> lat_dist(-90.0, 90.0);
    std::uniform_real_distribution<> lon_dist(-180.0, 180.0);
    
    for (int i = 0; i < num_to_generate; ++i) {
        double lat, pressure_anomaly, radius, speed;
        
        // Prefer certain latitudes for storm formation
        if (std::generate_canonical<double, 10>(gen) < 0.7) {  // 70% chance in mid-latitudes
            std::uniform_real_distribution<> mid_lat_dist(-60.0, 60.0);
            lat = mid_lat_dist(gen);
            
            if (std::abs(lat) < 30.0) {  // Tropical storms
                std::uniform_real_distribution<> trop_press_dist(-15.0, -5.0);
                std::uniform_real_distribution<> trop_radius_dist(300.0, 800.0);
                std::uniform_real_distribution<> trop_speed_dist(30.0, 60.0);
                pressure_anomaly = trop_press_dist(gen);
                radius = trop_radius_dist(gen);
                speed = trop_speed_dist(gen);
            } else {  // Extratropical cyclones
                std::uniform_real_distribution<> extra_press_dist(-25.0, -10.0);
                std::uniform_real_distribution<> extra_radius_dist(500.0, 1500.0);
                std::uniform_real_distribution<> extra_speed_dist(50.0, 100.0);
                pressure_anomaly = extra_press_dist(gen);
                radius = extra_radius_dist(gen);
                speed = extra_speed_dist(gen);
            }
        } else {  // 30% chance elsewhere
            lat = lat_dist(gen);
            std::uniform_real_distribution<> press_dist(-10.0, 10.0);
            std::uniform_real_distribution<> radius_dist(200.0, 1000.0);
            std::uniform_real_distribution<> speed_dist(20.0, 80.0);
            pressure_anomaly = press_dist(gen);
            radius = radius_dist(gen);
            speed = speed_dist(gen);
        }
        
        double lon = lon_dist(gen);
        
        // Find nearest vertex
        int nearest_vertex = 0;
        double min_dist = std::numeric_limits<double>::max();
        for (size_t i = 0; i < vertices.size(); ++i) {
            auto latlon = cartesian_to_lat_lon(vertices[i][0], vertices[i][1], vertices[i][2]);
            double dist = haversine_distance(lat, lon, latlon.first, latlon.second);
            if (dist < min_dist) {
                min_dist = dist;
                nearest_vertex = i;
            }
        }
        
        // Prefer forming over water
        if (elevations[nearest_vertex] < 0.0) {  // Water
            pressure_anomaly *= 1.5;  // Stronger over water
        }
        
        storms.emplace_back(lat, lon, pressure_anomaly, radius, speed);
    }
    
    return storms;
}